### PR TITLE
feat: support complex magical comment `import(/* webpackChunkName: "/dist/d/js_d" */ "./d")`

### PR DESCRIPTION
**Pull request by suxin2017**
## Related issue (if exists)

https://github.com/web-infra-dev/rspack/issues/2806
### <samp>🤖 Generated by Copilot at b6cb9e8</samp>

Improved the handling of webpack chunk names from magic comments in the JavaScript dependency scanner. Added a test case with input and output files to verify the functionality.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at b6cb9e8</samp>

*  Update the regex and logic for capturing the webpack chunk name from the magic comment ([link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-8281ab635aae318f63d8e3d743577a933a15bda2136698d2900e8bd8c711d2faL102-R102), [link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-8281ab635aae318f63d8e3d743577a933a15bda2136698d2900e8bd8c711d2faL115-R126))
*  Add a test case for the magic comment feature with different quote styles and chunk names ([link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-90a6cb45918dbd03fa7560798ab4eb28dc019f4cab94e02900ab977feff506e5R1-R8), [link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-a4d71534ec40d0197092cffa9d21b3cb06f51de1d18f53ed6e32d53b78bfa4b4L1-R1))
*  Add the expected output files for the test case, showing how the import statements are transformed into webpack chunk loading calls ([link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-896ce47440c79b9f60c65faac7b527f1186a7faf40e9a1c4bf9afdbab32bcc55L1-R12), [link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-4b5753c44aa52c6e6635baca51583b423d9fef0abb8df4f12aded550f072aef3L1-R5), [link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-4af81f590ad5b40f51057a77f8bb996a64e2e1e51058bb8cc334b5b387320ca5L1-R5), [link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-e389f4bf99842e84281afd7622010a539ace9abcfc446bdd2b0d40223510f163L1-R5))
*  Add the input files for the modules imported by the test case, containing simple console.log statements ([link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-10df113b6f640ae54696fbe18037a864d4db2e534a4b42f5dea5813ee5f9a63cL1-L-1), [link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-3d0eb6678536214c2eee865d3f169f71ae5ba5e50ea1d30b39c87df7ef30c0b1L1-L-1), [link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-d52ae299ddbfb384768c0e0d2ceba76f15ed6e84f3b8c6b3f552dd5d11c5bab5L1-L-1), [link](https://github.com/web-infra-dev/rspack/pull/2812/files?diff=unified&w=0#diff-a99ef256c0794b227b6257a94d3b0d6d57d2ece1f6413b87878a4535ecb17fdeL1-L-1))

</details>

### CODE REVIEW

File: crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs [hunk 1]
```diff
@@ -99,7 +99,7 @@ impl DependencyScanner<'_> {
     use once_cell::sync::Lazy;
     use swc_core::common::comments::CommentKind;
     static WEBPACK_CHUNK_NAME_CAPTURE_RE: Lazy<regex::Regex> = Lazy::new(|| {
-      regex::Regex::new(r#"webpackChunkName\s*:\s*["|'|`]\s*(\w[\w0-9]*)\s*["|'|`]"#)
+      regex::Regex::new(r#"webpackChunkName\s*:\s*"(?P<_1>(\./)?([\w0-9_\-]+/)*?[\w0-9_\-]+)"|'(?P<_2>(\./)?([\w0-9_\-]+/)*?[\w0-9_\-]+)'|`(?P<_3>(\./)?([\w0-9_\-]+/)*?[\w0-9_\-]+)`"#)
```
Comment (hyf0) [inline crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs:102]: I couldn't visualize the regex. [Link](https://regexper.com/#webpackChunkName%5Cs*%3A%5Cs*%22%28%3FP%3C_1%3E%28%5C.%2F%29%3F%28%5B%5Cw0-9_%5C-%5D%2B%2F%29*%3F%5B%5Cw0-9_%5C-%5D%2B%29%22%7C'%28%3FP%3C_2%3E%28%5C.%2F%29%3F%28%5B%5Cw0-9_%5C-%5D%2B%2F%29*%3F%5B%5Cw0-9_%5C-%5D%2B%29'%7C%60%28%3FP%3C_3%3E%28%5C.%2F%29%3F%28%5B%5Cw0-9_%5C-%5D%2B%2F%29*%3F%5B%5Cw0-9_%5C-%5D%2B%29%60)

Is this regex the only solution? If it's possible, I hope you could write a regex that could be visualize.

Comment (suxin2017) [reply crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs]: There are some special regex syntax that may be specific to rust (? P<name>expr)
https://docs.rs/regex/latest/regex/#grouping-and-flags

Below is a link to a visualization of this regular
https://regexper.com/#webpackChunkName%5Cs*%3A%5Cs*%28%22%28%28%5C.%5C%2F%29%3F%28%5B%5Cw0-9_%5C-%5D%2B%5C%2F%29*%3F%5B%5Cw0-9_%5C-%5D%2B%29%22%7C'%28%28%5C.%5C%2F%29%3F%28%5B%5Cw0-9_%5C-%5D%2B%5C%2F%29*%3F%5B%5Cw0-9_%5C-%5D%2B%29'%7C%60%28%28%5C.%5C%2F%29%3F%28%5B%5Cw0-9_%5C-%5D%2B%5C%2F%29*%3F%5B%5Cw0-9_%5C-%5D%2B%29%60%29

Comment (hyf0) [reply crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs]: Thanks. But I'm going to check out the usage and related tests of Webpack. I'm not sure we should support something(sub-folder) like `import(/* webpackChunkName: ./hello/elseo.js */, './foo.js')`.

Comment (suxin2017) [reply crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs]: webpack uses this approach
```
// regexp to match at least one "magic comment"
const webpackCommentRegExp = new RegExp(/(^|\W)webpack[A-Z]{1,}[A-Za-z]{1,}:/);

	const comments = this.getComments(range);
		if (comments.length === 0) {
			return EMPTY_COMMENT_OPTIONS;
		}
		let options = {};
		/** @type {unknown[]} */
		let errors = [];
		for (const comment of comments) {
			const { value } = comment;
                        // check is magin comment
			if (value && webpackCommentRegExp.test(value)) {
				// try compile only if webpack options comment is present
				try {
					for (let [key, val] of Object.entries(
                                                   // value = webpackChunkName: "xxx"
						vm.runInNewContext(`(function(){return {${value}};})()`)
					)) {
						if (typeof val === "object" && val !== null) {
							if (val.constructor.name === "RegExp") val = new RegExp(val);
							else val = JSON.parse(JSON.stringify(val));
						}
						options[key] = val;
					}
				} catch (e) {
					const newErr = new Error(String(e.message));
					newErr.stack = String(e.stack);
					Object.assign(newErr, { comment });
					errors.push(newErr);
				}
			}
		}
		return { options, errors };
```

Comment (suxin2017) [reply crates/rspack_plugin_javascript/src/visitors/dependency/scanner.rs]: > Thanks. But I'm going to check out the usage and related tests of Webpack. I'm not sure we should support something(sub-folder) like `import(/* webpackChunkName: ./hello/elseo.js */, './foo.js')`.

webpack doesn't support this so I don't think we need to either

**Review by hyf0 — APPROVED** [review]
Great work. Thank you!

**Comment by hyf0** [comment]
!canary

**Pull request merged**